applications/dashboard/controllers/api/MediaController.php

```diff
@@ -0,0 +1,306 @@
+use Vanilla\UploadedFileSchema;
+
+/**
+ * API Controller for `/media`.
+ */
+class MediaApiController extends AbstractApiController {
+
+    /** @var Schema */
+    private $idParamSchema;
+
+    /** @var MediaModel */
+    private $mediaModel;
+
+    /**
+     * MediaApiController constructor.
+     *
+     * @param MediaModel $mediaModel
+     */
+    public function __construct(MediaModel $mediaModel) {
+        $this->mediaModel = $mediaModel;
+    }
+
+    /**
+     * Delete a media item by ID.
+     *
+     * @param $id The media item's numeric ID.
+     */
+    public function delete($id) {
+        $this->permission('Garden.SignIn.Allow');
+
+        $in = $this->idParamSchema()->setDescription('Delete a media item.');
+        $out = $this->schema($this->fullSchema(), 'out');
+
+        $row = $this->mediaByID($id);
+        if ($row['InsertUserID'] !== $this->getSession()->UserID) {
+            $this->permission('Garden.Moderation.Manage');
+        }
+
+        $this->mediaModel->deleteID($id);
+    }
+
+    /**
+     * Delete a media item by its URL.
+     *
+     * @param array $query The request query.
+     */
+    public function delete_byUrl(array $query) {
+        $this->permission('Garden.SignIn.Allow');
+
+        $in = $this->schema(['url:s' => 'Full URL to the item.'], 'in')->setDescription('Delete a media item, using its URL.');
+        $out = $this->schema([], 'out');
+
+        $in->validate($query);
+
+        $row = $this->mediaByUrl($query['url']);
+        if ($row['InsertUserID'] !== $this->getSession()->UserID) {
+            $this->permission('Garden.Moderation.Manage');
+        }
+
+        $this->mediaModel->deleteID($row['MediaID']);
+    }
+
+    /**
+     * Process a user upload and insert into the media table.
+     *
+     * @param UploadedFile $upload An object representing an uploaded file.
+     * @param string $type The upload type (e.g. "image").
+     * @throws Exception if there was an error encountered when saving the upload.
+     * @return array
+     */
+    private function doUpload(UploadedFile $upload, $type) {
+        $file = $upload->getFile();
+
+        $media = [
+            'Name' => $upload->getClientFilename(),
+            'Type' => $upload->getClientMediaType(),
+            'Size' => $upload->getSize(),
+            'ForeignID' => $this->getSession()->UserID,
+            'ForeignTable' => 'embed'
+        ];
+
+        switch ($type) {
+            case 'image':
+                $imageSize = getimagesize($file);
+                if (is_array($imageSize)) {
+                    $media['ImageWidth'] = $imageSize[0];
+                    $media['ImageHeight'] = $imageSize[1];
+                }
+        }
+
+        $ext = pathinfo(strtolower($upload->getClientFilename()), PATHINFO_EXTENSION);
+        $destination = $this->generateUploadPath($ext, true);
+        $uploadResult = $this->saveUpload($upload, $destination);
+        $media['Path'] = $uploadResult['Name'];
```

Comment:
I think that the correct assignation would be `$media['Path'] = $uploadResult['SaveName'];` if I base myself on [this](https://github.com/vanilla/vanilla/blob/36faa6b09a6f8c95dee8ed888d9c240cbba0cc2a/plugins/editor/class.editor.plugin.php#L848)